manager.go

```diff
@@ -0,0 +1,150 @@
+package graphql
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GannettDigital/graphql/gqlerrors"
+	"github.com/GannettDigital/graphql/language/ast"
+)
+
+const workerStartDelay = 250 * time.Microsecond
+
+// TODO add a count of workers and a way to report that
+// TODO add a way to reclaim workers that are idle
+
+// completeRequest contains the information needed to complete a field.
+// A completeRequest is passed to a resolveManager worker which processes the request.
+// This is only used when completing multiple items concurrently such as needed for a list
+type completeRequest struct {
+	index    int
+	response chan<- completeResponse
+
+	eCtx       *executionContext
+	returnType Type
+	fieldASTs  []*ast.Field
+	info       ResolveInfo
+	value      interface{}
+}
+
+// completeResponse is the type containing the completion response which is sent over a channel as workers finish
+// processing.
+type completeResponse struct {
+	index  int
+	result interface{}
+}
+
+// resolveRequest contains the information needed to resolve a field.
+// A resolveRequest is passed to a resolveManager worker which processes the request.
+type resolveRequest struct {
+	fn       FieldResolveFn
+	name     string
+	params   ResolveParams
+	response chan<- resolverResponse
+}
+
+// resolverResponse is the type containing the resolver response which is sent over a channel as workers finish
+// processing.
+type resolverResponse struct {
+	err    error
+	name   string
+	result interface{}
+}
+
+// resolveManager runs resolve functions and completeValue requests in a set of worker go routines.
+// Having a set number of workers limits the churn of go routines while still providing parallel resolving of results
+// which is key to performance when some resolving requires a network call.
+//
+// The nature of the GraphQL resolving code is that a single resolve call could end up calling other resolve calls
+// as part of it. This means to avoid a full deadlock a hard limit on the number of workers can't be set, instead
+// a slight delay is added to give preference to reusing workers over creating new.
+type resolveManager struct {
+	completeRequests chan completeRequest
+	once             sync.Once
+	quit             chan bool
+	resolveRequests  chan resolveRequest
+}
+
+func newResolveManager(startingWorkers int) *resolveManager {
+	manager := &resolveManager{
+		completeRequests: make(chan completeRequest),
+		once:             sync.Once{},
+		quit:             make(chan bool),
+		resolveRequests:  make(chan resolveRequest),
+	}
+
+	for i := 0; i < startingWorkers; i++ {
+		go manager.newWorker()
+	}
+
+	return manager
+}
+
+func (manager *resolveManager) completeRequest(req completeRequest) {
+	for {
+		select {
+		case manager.completeRequests <- req:
+			return
+		case <-time.After(workerStartDelay):
+			go manager.newWorker()
```

Comment:
Hypothetically this would keep running forever, do we want to scale back down after handling the requests at hand?

Reply:
Yes I think so, I have TODO item in the manager code to implement that.